### DISCUSSION TITLE
Add header and footer to specific page and resize current content

### DISCUSSION THREAD

**Discussion by krzysiekkajot** [qpdf/qpdf#1256] [Q&A] [answered]
Hi, I'm new with qpdf. I suppose to write a small cmld line tool for adding user defined texts on the top and bottom on user specified page. How can I do this with qpdf?
I was abble to get working example which will open file, take specific page, getKey with "/Contents" and then convert content to Array, and add at the end new stream with data like "0 0 0 RG\n2 w\n50 250 m\n50 500 l\nS\n" and then replace the orginal "/Contents" With new array and save pdf.
Is it a proper solution? How to add the text? Should I use similar "magic string" (is there any helper page with them).
In case of current content is already at position where I'm trying to add text how to scale the current page to make it smaller to make space for my additional lines?

BR
  Krzysztof

**Comment by jberkenbilt** [comment, marked as answer]
The attached file is a Python script I wrote for my own personal use. It uses Pikepdf, which is Python wrapper around qpdf, but you can see what the code is doing and do the same thing with the C++ API. Hopefully this will give you a good idea of how to do it. In addition to adding content code after the original page, this sample wraps the original content with `q` and `Q` to protect you from matrix changes within the content stream. It also uses one of the built-in fonts to avoid font embedding, though this restricts it to characters in the PDF Doc Encoding character set, which is more or less ISO-8859-1 with a handful of extras.

[pdf-add-text.txt](https://github.com/user-attachments/files/16432410/pdf-add-text.txt)

**Reply by jberkenbilt** [reply]
sorry about the .txt extension. GitHub has some restrictions on file names for attachments.

**Reply by krzysiekkajot** [reply]
BIG THANKS for example! It was great. It take me some time to figure out I need to add two streams. In my C++ app I just create new content with new lines and then at it at the end of existing content, but it didn't work. Why it need additional stream at the beginning containing only "q\n"? This is kind of pdf format specific case? Can you advice any source/reference from I can learn when I need to add this q thing and when not?

I also need to scale current content in case it intersects with lines I want to add. Is there a possibility I can somehow pack existing content in some object/stream and then use some scale factor for everything what is in that object/stream? Or maybe I need to iterate whole existing page content and then change its size and position or apply some scale factor? Do you know any example? Currently I'm trying to iterate but I don't have any idea how to get x,y and size in case of page content item is stream.

Thank you in advance.

**Reply by krzysiekkajot** [reply]
ok, it looks like pikepdf help page is giving some explanations, I will try them.

**Reply by jberkenbilt** [reply]
Very briefly: `q` and `Q` surround an arbitrary block of content in such a way that certain aspects of the state are preserved. `q` saves the state, and `Q` restores it. One part of the state that is preserved is the transformation matrix. The reason for the `q` before and `Q` after is to make sure that, whatever code exists in the page, you are protected from it for your addition. If you wanted to resize the page or something, you could simply add your own `cm` operation to scale/translate the page after the `q`.

This solution is still not perfect in that it doesn't protect you from certain errors in the original page. For example, if the original page itself contains unbalanced `q`/`Q` operators, then your addition may still be influenced by code in the page. To get around this, you can convert the page to a form XObject and create a brand new page that displays the original page as a form XObject and then adds your additional content after. This should basically always work. The qpdf library has a method in `QPDFPageObjectHelper` called `getFormXobjectForPage`. The `examples/pdf-overlay-page.cc` example in the qpdf source tree (and possibly installed in /usr/share/doc/qpdf-dev/examples on debian-based systems) shows you how to do this. Perhaps there's stuff in the pikepdf docs as well. There's also `QPDFPageObjectHelper::placeFormXObject` that does some of the mechanics of placing the form xobject.

So, in summary, the most robust way to do this is to create a form XObject out of the original page, place it on the new page in a manner that preserves its scale or alters it if you want, and then add your new code after it. `QPDFPageObjectHelper`'s `getFormXObjectForPage` provides a low-level interface to help with this, and `placeFormXObject` may provide a higher-level interface. It's a bit complex to know how it all works, but in the end, it should just be a few lines of code. My example predates the form XObject stuff in qpdf and doesn't use that more robust approach.

I've been a bit vague here in the interest of time, but let me know if you need further guidance after giving this a look.

**Reply by krzysiekkajot** [reply]
GREAT! big thanks

I found also this: https://medium.com/@jberkenbilt/examining-a-pdf-file-with-qpdf-219ff19d5c47 [1] and then https://medium.com/@jberkenbilt/the-structure-of-a-pdf-file-6f08114a58f6 [2]. I'm going through them.

It is still messy for me how to decode
<MAGIC NUMBERS> Tj
It is not base64, I found that somhere in [1] it was mentioned that this binary magic contain the glyph number from font (need to read it again) - can you provide some hint how to decode <MAGIC STRING> Tj? When I was playing with python I saw that sometimes it was <MAGIC> Tj but sometimes PLAIN_TEXT Tj. Can it be somehow decode when using json output or qdf output?

**Reply by krzysiekkajot** [reply]
I think last question in this subject, to be sure I on good path:

```
%% Contents for page 1
%% Original object ID: 5 0
11 0 obj
<<
  /Length 12 0 R
>>
stream
q
q
BT
0 0 0 rg
/GS0 gs
/F0 8 Tf                          <- use font F0
555.7299 37.5815 Td <03ed> Tj     <- draw glyph 03ed at position specified by Td
4.0547 0 Td <036c> Tj             <- draw glyph 036c at position specified by Td
3.0898 0 Td <03f1> Tj
/F1 9 Tf                          <- use font F1
```

So I need to lookup in font table/stream/resource for Fx and then decode the text? I'm bit curious why pikepdf is giving me nicely decoded strings but sometimes, for different pfds gives me kind of noise. I'm using:

```
    instructions = pikepdf.parse_content_stream(page)
    data = pikepdf.unparse_content_stream(instructions)
    print(data.decode('ascii'))
```

As I understand it is kind of decoding issue ascii/utf8 or maybe even Windows-1250 or ISO 8859-2 for some documents in my case. How to find which one was used for specific string or maybe whole document?

**Reply by jberkenbilt** [reply]
Welcome to the messy world of text extraction and decoding. This is something else I should blog about someday. I just added it to my blog backlog, though who knows how long it will take to get to it. :-)

In a nutshell (slightly simplified from reality), a font in PDF consists of a font or font subset, which is just a part of a font containing only some of the glyphs, and optionally, a map indicating the corresponding Unicode character attached to each position in the font. You can think of the font as being an array of glyphs, and the code point value in the string passed to Tj as an index into that array. The PDF spec allows for strings to be encoded in various ways. Unicode strings are allowed in structural places in the file, like metadata and outlines (bookmarks), but not in content. Content is typically encoded with PDF Doc encoding, but you can also choose "standard encoding", which is ASCII, WinAnsiEncoding, and MacRomanEncoding. The PDF spec includes a table that maps these to and from Unicode. They are all 7-bit or 8-bit encodings and as such can't represent characters in most fonts.

Most of the time, a font subset uses an 8-bit index, which is fine as long as no more than 256 characters from the font are used. Sometimes they have a 16-bit index. PDF files usually contain font subsets. It's common for a PDF file to contain characters from multiple fonts, and frequently, only a few characters from a font are used. A very common pattern is for the first character that appears to be codepoint 0, the next one 1, the next one 2, etc. So if you had a file that had a specific font, and the only word in the file that used that font was `potato`, you might have a font subset containing only p, o, t, and a in that order with 0 = p, 1 = o, 2 = t, and 3 = a, and the string passed to Tj would look like <000102030201>. If you want to convert this back to text, you have to see what font is used and look up its map. The font is whatever is referenced in the most recent Tf call. You can map the name of the font (e.g. /F0, /F1, etc.) to the font dictionary in the page's Resource dictionary.

If you're lucky, the font information includes a `/ToUnicode` field which maps ranges of codepoints to ranges of Unicode characters. In this case, you can extract the text. If you're not lucky, `/ToUnicode` is absent, and you're out of luck. Most readers will fall back to interpreting the codepoints as Unicode, and you'll get garbage. At this point, the only way to really extract the text is to do something like OCR. In principal, I imagine one could build something that recognizes the glyphs individually by having a known database of fonts, doing some OCR-like process on the rendered vector images or the vector images themselves, looking at the glyphs in context and solving it like a letter scramble puzzle, or some combination of the above. With an adequate amount of text, you could probably do it perfectly, but obviously something like this goes way beyond qpdf or a simple tool. Also, there's nothing to stop someone from creating a font full of glyphs that don't map to Unicode characters at all...maybe you've designed a special board game and have created a font with glyphs corresponding to symbols used in the game or something. If that weren't enough operators like Tj are rendering symbols with spacing, and if the typesetting was done by the tool that generated the PDF, it's extremely common for the fragments of text to not correspond to word boundaries. Then the text extraction process has to guess the word boundaries by looking at the relative spacing between the letters. Of course, PDF files can contain images of text as well, and it's not that unusual to encounter hybrid files where people have, for example, added native text on top of a scanned form or added edits to a file so that bits of text that are proximate on the rendered page may not be anywhere near each other in the PDF content code.

Bottom line: a PDF file that has been designed for text extraction using accessibility features can have its text extracted perfectly. A file that uses non-subsetted fonts or built-in fonts may have character values in the PDF that map to the actual character codepoints in the base encoding, in which case the text can be extracted. A file that uses font subsets or characters that fall outside of one of the built-in PDF encoding systems may include a `/ToUnicode` map. If so, you can use that to get the text. If none of the above, then you can't get the text without some kind of character recognition.

I hope that explanation is useful. I will probably use some of this explanation in my blog if I ever write it.

**Reply by krzysiekkajot** [reply]
Wow, you are really PDF PhD master :) I've made fast pdf -> qdf conversion then analyze few text and transforming them using ToUnicode maps (my pdf's have them!) and it works! Now need to implement it in C++ using qpdf lib, Big thanks again!

BTW bad news from me is that pdf contains only some glyphs, so not all characters, in some moment in the future I need to add new few lines to pdf and it looks like it could be a problem if there will be characters which not exists in pdf font glyphs :/ But as I understand in such kind of case I can take the glyphs from some ttf or other font file and add it updating glyphs and ToUnicode structs. I found that glyphs should be in /FontFile2 file and it looks like *.ttf (CheckFileType.com report it like that).

Thanks again!

**Reply by jberkenbilt** [reply]
Well, I actually wrote the blog. If you want a real deep dive, you can find it at https://medium.com/@jberkenbilt/text-in-pdf-introduction-df3dd3dfa9ea, which is the first in a five-part series. All five parts are published.

**Reply by krzysiekkajot** [reply]
I read it for sure! :) THANKS
